File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/Source.java
package com.ficture7.aasexplorer.model;

import com.ficture7.aasexplorer.client.Client;

import java.net.URI;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;

import static com.ficture7.aasexplorer.util.ObjectUtil.checkNotNull;

/**
 * Represents a source of a content.
 *
 * @author FICTURE7
 */
public interface Source {

    /**
     * Returns the {@link Client} instance which acquired the source.
     *
     * @return {@link Client} instance which acquired the source.
     */
    Client client();

    /**
     * Returns the {@link Date} when the source was acquired.
     *
     * @return {@link Date} when the source was acquired.
     */
    Date date();

    /**
     * Returns the {@link URI} of the source.
     *
     * @return {@link URI} of the source.
     */
    URI uri();

    /**
     * Represents a list of {@link Source} instances.
     *
     * @param <T> Type of {@link Source}.
     */
    class List<T extends Source> implements Iterable<T> {

        private final java.util.List<T> sources;

        /**
         * Constructs a new instance of the {@link List} class.
         */
        public List() {
            sources = new ArrayList<>();
        }

        /**
         * Returns the number of {@link Source} in the {@link List}.
         *
         * @return Number of {@link Source} in the {@link List}.
         */
        public int size() {
            return sources.size();
        }

        /**
         * Adds the specified {@link Source} instance to the {@link List}.
         *
         * @param source {@link Source} instance to add.
         * @throws NullPointerException {@code source} is null.
         */
        public void add(T source) {
            checkNotNull(source, "source");

            for (T s : sources) {
                if (s.client().getClass() == source.client().getClass()) {
                    throw new IllegalStateException("Already contains a Source instance with the same Client type.");
                }
            }

            sources.add(source);
        }

        /**
         * Removes the specified {@link Source} instance from the {@link List}.
         *
         * @param source {@link Source} instance to remove.
         * @throws NullPointerException {@code source} is null.
         */
        public void remove(T source) {
            checkNotNull(source, "source");

            sources.remove(source);
        }

        /**
         * Clears the {@link List}.
         */
        public void clear() {
            sources.clear();
        }

        /**
         * Returns an iterator over the elements of the {@link List}.
         * @return An iterator over the elements of the {@link List}.
         */
        @Override
        public Iterator<T> iterator() {
            return sources.iterator();
        }
    }
}

File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/Source.kt
package com.ficture7.aasexplorer.model

import com.ficture7.aasexplorer.client.Client

import java.net.URI
import java.util.ArrayList
import java.util.Date

/**
 * Represents a source of a content.
 *
 * @author FICTURE7
 */
interface Source {

    /**
     * Returns the [Client] instance which acquired the source.
     *
     * @return [Client] instance which acquired the source.
     */
    val client: Client

    /**
     * Returns the [Date] when the source was acquired.
     *
     * @return [Date] when the source was acquired.
     */
    val date: Date

    /**
     * Returns the [URI] of the source.
     *
     * @return [URI] of the source.
     */
    val URI: URI

    /**
     * Represents a list of [Source] instances.
     *
     * @param <T> Type of [Source].
     */
    class List<T : Source> : Iterable<T> {

        private val sources: MutableList<T>

        /**
         * Constructs a new instance of the [List] class.
         */
        constructor() {
            sources = ArrayList()
        }

        /**
         * Returns the getNumber of [Source] in the [List].
         *
         * @return Number of [Source] in the [List].
         */
        fun size(): Int {
            return sources.size
        }

        /**
         * Adds the specified [Source] instance to the [List].
         *
         * @param source [Source] instance to add.
         */
        fun add(source: T) {
            for (s in sources) {
                if (s.client.javaClass == source.client.javaClass) {
                    throw IllegalStateException("Already contains a Source instance with the same Client type.")
                }
            }

            sources.add(source)
        }

        /**
         * Removes the specified [Source] instance from the [List].
         *
         * @param source [Source] instance to remove.
         */
        fun remove(source: T) {
            sources.remove(source)
        }

        /**
         * Clears the [List].
         */
        fun clear() {
            sources.clear()
        }

        /**
         * Returns an iterator over the elements of the [List].
         * @return An iterator over the elements of the [List].
         */
        override fun iterator(): Iterator<T> {
            return sources.iterator()
        }
    }
}
